src/frontend/src/utils/countdown.ts

```diff
@@ -0,0 +1,78 @@
+import { render } from "lit-html";
+import { displayError } from "../components/displayError";
+
+export class Countdown {
+  private timeoutHandle;
+  private expected;
+  private readonly endTimestamp;
+
+  constructor(
+    private updateFunc: () => void,
+    private interval: number,
+    private timeoutFunc: () => void,
+    endTimestamp: bigint
+  ) {
+    this.expected = Date.now() + this.interval;
+    this.timeoutHandle = window.setTimeout(() => this.step(), this.interval);
+    this.endTimestamp = Number(endTimestamp / BigInt("1000000"));
+  }
+
+  public stop(): void {
+    window.clearTimeout(this.timeoutHandle);
+  }
+
+  private async step() {
+    const now = Date.now();
+    if (now >= this.endTimestamp) {
+      await this.timeoutFunc();
+      return;
+    }
+    const drift = now - this.expected;
+    this.updateFunc();
+    this.expected += this.interval;
+    this.timeoutHandle = window.setTimeout(
+      () => this.step(),
+      Math.max(0, this.interval - drift)
+    );
+  }
+}
+
+export const setupCountdown = (
+  timestamp: bigint,
+  continueFunc: () => Promise<void>
+): Countdown => {
+  const timer = document.getElementById("timer") as HTMLElement;
```

Comment:
How about passing the element as an argument instead? Using global IDs like this is really brittle/dangerous, especially given the amount of usage doc in this file :)))

Alternatively, a standalone [component](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements) might encapsulate the logic better